lib/bolt/transpiler.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+require 'bolt/pal/yaml_plan'
+require 'bolt/pal/yaml_plan/loader'
+require 'pathname'
+
+module Bolt
+  class YamlTranspiler
+    def self.transpile(plan_path)
+      # TODO: make path absolute
+      unless Pathname.new(plan_path).absolute?
+
+      end
+
+      @plan_path = plan_path
+      @filename, @module = parse_path
+      validate
+
+      plan_object = parse_plan
+
+      plan_string = String.new("plan #{plan_object.name}(")
+      # Parameters are Bolt::PAL::YamlPlan::Parameter
+      plan_object.parameters&.each_with_index do |param, i|
+        # Param type
+        type = if param.type_expr.respond_to?(:type_string)
+                 param.type_expr.type_string
+               else
+                 param.type_expr.to_s
+               end
+        plan_string << "\n\s\s"
+        plan_string << type.to_s if type
+
+        # Param name
+        plan_string << "$#{param.name}"
+
+        # Param default
+        if param.value
+          default = type =~ /String/ ? "'#{param.value}'" : param.value
+          plan_string << " = #{default}"
+        end
+
+        # If it's the last parameter add a newline and no comma
+        last = i + 1 == plan_object.parameters.length ? "\n" : ","
+        # This encodes strangely if we << directly to plan_string
+        plan_string << last
+      end
+      plan_string << ") {\n"
+
+      plan_object.steps&.each do |step|
+        step.transform_values! { |v| plan_object.stringify(v) }
+        plan_string << "$#{step['name']} = " if step['name']
+        step_type, *extra_keys = Bolt::PAL::YamlPlan::Evaluator::STEP_KEYS.select { |key| step.key?(key) }
+
+        if !step_type || extra_keys.any?
+          raise Bolt::ConvertError.new("Can't convert unsupported step type #{step.name}", @plan_path)
+        end
+
+        target = step['target']
+        description = step.fetch('description', nil)
+        parameters = step.fetch('parameters', {})
+        if step_type == 'script' && step.key?('arguments')
+          parameters['arguments'] = step['arguments']
+        end
+
+        case step_type
+        when 'command', 'task', 'script', 'plan'
```

Comment:
In general I think an erb template might make this easier than constructing one giant string.